### PR TITLE
Add Gradle integration tests [WIP]

### LINKED ISSUE #431: Create integration tests that resolve artifacts from Strongbox using Gradle
We need to add integration tests (in the [strongbox-web-integration-tests](https://github.com/strongbox/strongbox-web-integration-tests/) project) that resolve and deploy artifacts using Gradle.

There's a ["Hello, World!"](https://github.com/strongbox/strongbox-examples/tree/master/hello-strongbox-gradle/) application that we've already created for illustrating how to configure Gradle to work with Strongbox.

The tests should be invoked in a similar way to how they're currently being invoked for NuGet -- using Groovy scripts. There should be a Maven profile similar to the one for NuGet (check the one with`<id>run-nuget-it-tests</id>` for an example).

The test case must cover at least the following:
- Resolve dependencies
- Deploy an artifact

Please, note that the work for this will have to be done in a different project -- [strongbox-web-integration-tests](https://github.com/strongbox/strongbox-web-integration-tests/).

This issue relates to #432 and #430.

### PR DESCRIPTION
**Pull request by danielbrzn**
Aims to resolve [strongbox#431](https://github.com/strongbox/strongbox/issues/431).

I've managed to add in artifact deployment using Gradle, but I'm not particularly sure on these following aspects of implementing the other tests. 

- How can I check that the artifact is indeed deployed on the local Strongbox server?
- How can I implement a test for resolving dependencies? (Would this entail having an artifact already deployed on Strongbox and then using gradle to resolve the dependencies from Strongbox?)

It's also my first time using Groovy and I'd appreciate comments on whether or not I'm heading in the correct direction with these tests.

### CODE REVIEW

File: src/gradle-it/build.gradle [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+description = "Example project that shows how to use gradle to build and deploy artifacts to Strongbox"
+
+ext.username = System.getProperty('credentials.username') != null ? System.getProperty('credentials.username') : mavenUser
+ext.password = System.getProperty('credentials.password') != null ? System.getProperty('credentials.password') : mavenPassword
+
+configurations {
+    deployerJars
+}
+
+repositories {
```
Comment (carlspring) [inline src/gradle-it/build.gradle:19]: Here you need just one repository:
```
maven { url "http://localhost:48080/storages/public/public-group/" }
```

Basically, you should deploy artifacts to the hosted repositories and resolve them via the group. 

Quick rule of thumb:
* Hosted repositories (literally) host artifacts. You can deploy to them.
* Proxy repositories can only cache requested artifacts after resolving them from a remote. You cannot deploy to them.
* Group repositories are used to serve a list of repositories under the same URL. Artifacts are resolved based on the order in which repositories are listed in the group. You cannot deploy to group repositories.

Please, drop these:
```
    maven { url "http://localhost:48080/storages/storage0/releases/" }
    maven { url "http://localhost:48080/storages/storage0/snapshots/" }
```

Also, the repository I showed you, contains Maven Central, which is where not only Maven, but also tools like Gradle, SBT, Ant+Ivy, Leiningen and so on resolve their dependencies from. (So, in this case you'll be covered).

**Comment by carlspring** [comment]
> I've managed to add in artifact deployment using Gradle, but I'm not particularly sure on these following aspects of implementing the other tests. 
> 
> - How can I check that the artifact is indeed deployed on the local Strongbox server?

Well, if you're deploying a `1.0-SNAPSHOT` version to the `snapshots` repository, you could check for files in the following directory;
```
strongbox-vault/storages/storage0/snapshots/org/carlspring/strongbox/examples/1.0-SNAPSHOT/
```
Obviously, you should have the jar there.

> - How can I implement a test for resolving dependencies? (Would this entail having an artifact already deployed on Strongbox and then using gradle to resolve the dependencies from Strongbox?)

One way is to add a dependency to some third-party library like `ch.qos.logback;logback-core:1.2.3` and check that it has been resolved, in a similar way as illustrated above. (You can check manually using `find` through which proxy repository the artifact was downloaded and then add a check in your test to check this programatically instead).

> It's also my first time using Groovy and I'd appreciate comments on whether or not I'm heading in the correct direction with these tests.

So far, so good! Keep it up! :)

This pull looks along the lines of what we need, thanks! I have some minor remarks.

I won't merge it before we have our Jenkins slaves set up to have Gradle, though (it may be a few more days).

**Comment by danielbrzn** [comment]
@carlspring 

Managed to clear up the artifact deployment test, so i'm working on resolving dependencies. 

Would the dependency be added to `build.gradle` or is there some other file that I must add it to? What do you mean by checking manually using `find`?

Am I understanding the flow right:
Strongbox resolves the dependency from a remote, then Gradle resolves the dependency using Strongbox's proxy repo, we do a check as to whether the artifact was actually resolved by checking for the downloaded artifact.

**Comment by carlspring** [comment]
You need to define the dependency in the `build.gradle` file.

And, yes, yes are right.

File: src/gradle-it/test-gradle-common-flow.groovy [hunk 2]
```diff
@@ -5,34 +5,36 @@ this.metaClass.mixin baseScript
...
 
-runCommand(gradlePath, String.format("$gradleExec clean upload -Dcredentials.username=%s -Dcredentials.password=%s -info",
+runCommand(gradlePath, String.format("$gradleExec upload -Dcredentials.username=%s -Dcredentials.password=%s",
                                     username,
                                     password))
 
 // check if dependency was resolved from Strongbox
-assert targetPath.resolve('strongbox-vault\\storages\\storage-common-proxies\\carlspring\\com\\fasterxml\\jackson\\core\\' +
-                          '\\jackson-databind\\2.9.4\\').resolve('jackson-databind-2.9.4.jar').toFile().exists();
+assert targetPath.resolve('strongbox-vault/storages/storage-common-proxies/carlspring/com/fasterxml/jackson/core/' +
```
Comment (SoulSilence) [inline src/gradle-it/test-gradle-common-flow.groovy]: I have one question. Is this change necessary? I mean change one slash by another. If it has sense maybe it's better to write it in os independent way like targetPath.resolve("strongbox-vault").resolve("storages")...

Comment (carlspring) [reply src/gradle-it/test-gradle-common-flow.groovy]: Yes, I specifically requested it, as it doesn't work under Linux. :)

**Comment by carlspring** [comment]
Thanks! :)
